### PR TITLE
infoschema, executor: Support getting current executing SQL text in TIDB_TRX table

### PR DESCRIPTION
**Pull request by MyonKeminta**
### What problem does this PR solve?

Problem Summary: Currently tables in Lock View shows SQL digests, which is difficult for user to get the corresponding SQL text. It's better to make easier to get the SQL text, which provides better user experience.

### What is changed and how it works?

What's Changed: This PR adds a new column `CURRENT_SQL_DIGEST_TEXT` to `TIDB_TRX` and `CLUSTER_TIDB_TRX` table that shows the SQL text corresponding to `CURRENT_SQL_DIGEST`. This PR also added utilities for getting SQL text by digest, which can be reused for other tables.

### Check List 

Tests 

- [ ] Unit test (TODO)

Side effects

- [x] Performance regression: Consumes more CPU
- [x] Performance regression: Consumes more Memory

Documentation

- [x] Affects user behaviors
  
### Release note 

```release-note
Add SQL text information to `INFORMATION_SCHEMA.TIDB_TRX` and `INFORMATION_SCHEMA.CLUSTER_TIDB_TRX` table
```

### CODE REVIEW

**Comment by MyonKeminta** [comment]
@cfzjywxk @longfangsong Do you have time to take a look?

File: executor/infoschema_reader.go [hunk 1]
```diff
@@ -2167,6 +2135,109 @@ func (e *stmtSummaryTableRetriever) retrieve(ctx context.Context, sctx sessionct
...
+			}
+		}
+
+		res = make([][]types.Datum, 0, end-start)
+
+		for i := start; i < end; i++ {
+			row := make([]types.Datum, 0, len(e.columns))
+			for _, c := range e.columns {
+				if c.Name.O == util.ClusterTableInstanceColumnName {
+					row = append(row, types.NewDatum(instanceAddr))
```
Comment (cfzjywxk) [inline executor/infoschema_reader.go:2217]: Is it expected to leave an empty string if the table is not `CLUSTER_TIDB_TRX`?

File: executor/infoschema_reader.go [hunk 2]
```diff
@@ -2167,6 +2135,109 @@ func (e *stmtSummaryTableRetriever) retrieve(ctx context.Context, sctx sessionct
 	return rows, nil
 }
 
+type tidbTrxTableRetriever struct {
```
Comment (cfzjywxk) [inline executor/infoschema_reader.go:2138]: We could add some comments for this new struct.

File: executor/utils.go [hunk 3]
```diff
@@ -53,3 +58,201 @@ func deleteFromSet(set []string, value string) []string {
 	}
 	return set
 }
+
+// SQLDigestTextRetriever is used to find the normalized SQL statement text by SQL digests in statements_summary table.
+// It's exported for text purposes.
```
Comment (cfzjywxk) [inline executor/utils.go:63]: `test porposes`?

File: executor/utils.go [hunk 4]
```diff
@@ -53,3 +58,201 @@ func deleteFromSet(set []string, value string) []string {
...
+		}
+	}
+	return res, nil
+}
+
+// runQuery runs query to the system tables to fetch normalized SQL texts corresponding to given SQL digests, if
+// `inValues` is given, or all digest -> SQL mappings otherwise. If `queryGlobal` is false, it queries
+// information_schema.statements_summary and information_schema.statements_summary_history; otherwise, it queries the
+// cluster version of these two tables.
+func (r *SQLDigestTextRetriever) runQuery(ctx context.Context, sctx sessionctx.Context, queryGlobal bool, inValues []interface{}) (map[string]string, error) {
```
Comment (cfzjywxk) [inline executor/utils.go:103]: How about `runFetchDigestQuery`, `runQuery` seems to be a general query executor.

File: executor/utils.go [hunk 5]
```diff
@@ -53,3 +58,201 @@ func deleteFromSet(set []string, value string) []string {
...
+	for _, digest := range inValues {
+		if text, ok := data[digest.(string)]; ok {
+			res[digest.(string)] = text
+		}
+	}
+	return res, nil
+}
+
+// runQuery runs query to the system tables to fetch normalized SQL texts corresponding to given SQL digests, if
+// `inValues` is given, or all digest -> SQL mappings otherwise. If `queryGlobal` is false, it queries
```
Comment (cfzjywxk) [inline executor/utils.go:100]: How about `it returns the kv mapping of sql digests and sql texts`?

File: executor/utils.go [hunk 6]
```diff
@@ -53,3 +58,201 @@ func deleteFromSet(set []string, value string) []string {
...
+	}
+
+	res := make(map[string]string, len(rows))
+	for _, row := range rows {
+		res[row.GetString(0)] = row.GetString(1)
+	}
+	return res, nil
+}
+
+func (r *SQLDigestTextRetriever) updateWithQueryResult(queryResult map[string]string) {
```
Comment (cfzjywxk) [inline executor/utils.go:145]: How about `updateDigestInfo`. Or we could add some comments about the targets of the update.

File: executor/utils.go [hunk 7]
```diff
@@ -53,3 +58,201 @@ func deleteFromSet(set []string, value string) []string {
...
+	}
+
+	var queryResult map[string]string
+	if len(r.SQLDigestsMap) <= r.fetchAllLimit {
+		queryResult, err = r.runQuery(ctx, sctx, true, unknownDigests)
+		if err != nil {
+			return errors.Trace(err)
+		}
+	} else {
+		queryResult, err = r.runQuery(ctx, sctx, true, nil)
```
Comment (cfzjywxk) [inline executor/utils.go:217]: Seems no result number limit here? Could it consumes much memory?

File: executor/infoschema_reader.go (hunk 1 above)
Comment (MyonKeminta) [reply executor/infoschema_reader.go:2217]: This function handles `CLUSTER_TIDB_TRX` and `TIDB_TRX`, and the `INSTANCE` column only exists in cluster tables.

File: executor/utils.go (hunk 3 above)
Comment (MyonKeminta) [reply executor/utils.go:63]: test purposes 😂

File: executor/utils.go (hunk 7 above)
Comment (MyonKeminta) [reply executor/utils.go:217]: I asked @crazycs520 before and he told me the statements_summary_history is unlikely to cause OOM in real situations, but I'm afraid so. Currently I didn't limit it for simplicity. Do you have any idea about this? Should we always use the `WHERE  IN` clause to filter the result?

Comment (cfzjywxk) [reply executor/utils.go:217]: Theoretically it's better to use limit and where clause in queries,  if the history results will not be too big I think it's ok to keep the implementation here to make it simple.

**Comment by MyonKeminta** [comment]
@cfzjywxk PTAL again, thanks!

**Comment by MyonKeminta** [comment]
Wait.. seems something is wrong about privileges..

**Review by cfzjywxk — APPROVED** [review]
/lgtm

**Review by youjiali1995 — APPROVED** [review]
LGTM

**Pull request merged**